Cubical/Data/DiffInt/Properties.agda

```diff
@@ -52,3 +58,238 @@ private
+
+bwd : ℤ -> Int
+bwd = elim {R = rel} {B = λ _ → Int} (λ _ → Int.isSetInt) bwd' bwd'-respects-rel
+
+bwd-fwd : ∀ (x : Int) -> bwd (fwd x) ≡ x
+bwd-fwd (Int.pos zero) = refl
+bwd-fwd (Int.pos (suc n)) i = sucInt (bwd-fwd (Int.pos n) i)
+bwd-fwd (Int.negsuc n) = refl
+
+fwd-bwd' : (a : ℕ × ℕ) → fwd (bwd [ a ]) ≡ [ a ]
+fwd-bwd' a@(zero , zero) = refl
+fwd-bwd' a@(zero , suc a⁻) = refl
+fwd-bwd' a@(suc a⁺ , a⁻) = sym (suc-⟦⟧ (bwd [ a⁺ , a⁻ ])) ∙ (λ i → sucℤ (fwd-bwd' (a⁺ , a⁻) i))
+
+fwd-bwd : ∀ (z : ℤ) -> fwd (bwd z) ≡ z
+fwd-bwd = elimProp {R = rel} (λ _  → isSetℤ _ _) fwd-bwd'
+
+Int≡ℤ : Int ≡ ℤ
+Int≡ℤ = isoToPath (iso fwd bwd fwd-bwd bwd-fwd)
+
+infix  8 -_
+infixl 7 _·_
+infixl 6 _+_
+
+_+'_ : (a b : ℕ × ℕ) → ℤ
+(a⁺ , a⁻) +' (b⁺ , b⁻) = [ a⁺ +ⁿ b⁺ , a⁻ +ⁿ b⁻ ]
+
+private
+  commˡⁿ : ∀ a b c → a +ⁿ b +ⁿ c ≡ a +ⁿ c +ⁿ b
+  commˡⁿ a b c = sym (ℕ.+-assoc a b c) ∙ (λ i → a +ⁿ ℕ.+-comm b c i) ∙ ℕ.+-assoc a c b
+
+  commʳⁿ : ∀ a b c → a +ⁿ (b +ⁿ c) ≡ b +ⁿ (a +ⁿ c)
+  commʳⁿ a b c = (ℕ.+-assoc a b c) ∙ (λ i → ℕ.+-comm a b i +ⁿ c) ∙ sym (ℕ.+-assoc b a c)
+
+  lem0 : ∀ a b c d → (a +ⁿ b) +ⁿ (c +ⁿ d) ≡ (a +ⁿ c) +ⁿ (b +ⁿ d)
+  lem0 a b c d = ℕ.+-assoc (a +ⁿ b) c d ∙ (λ i → commˡⁿ a b c i +ⁿ d) ∙ sym (ℕ.+-assoc (a +ⁿ c) b d)
+
++ⁿ-creates-rel-≡ : ∀ a⁺ a⁻ x → _≡_ {A = ℤ} [ a⁺ , a⁻ ] [ a⁺ +ⁿ x , a⁻ +ⁿ x ]
++ⁿ-creates-rel-≡ a⁺ a⁻ x = eq/ (a⁺ , a⁻) (a⁺ +ⁿ x , a⁻ +ⁿ x) ((λ i → a⁺ +ⁿ ℕ.+-comm a⁻ x i) ∙ ℕ.+-assoc a⁺ x a⁻)
+
++-respects-relʳ : (a b c : ℕ × ℕ) → rel a b → (a +' c) ≡ (b +' c)
++-respects-relʳ a@(a⁺ , a⁻) b@(b⁺ , b⁻) c@(c⁺ , c⁻) p = eq/ {R = rel} (a⁺ +ⁿ c⁺ , a⁻ +ⁿ c⁻) (b⁺ +ⁿ c⁺ , b⁻ +ⁿ c⁻) (
+  (a⁺ +ⁿ c⁺) +ⁿ (b⁻ +ⁿ c⁻) ≡⟨ lem0 a⁺ c⁺ b⁻ c⁻ ⟩
+  (a⁺ +ⁿ b⁻) +ⁿ (c⁺ +ⁿ c⁻) ≡[ i ]⟨ p i +ⁿ (c⁺ +ⁿ c⁻) ⟩
+  (b⁺ +ⁿ a⁻) +ⁿ (c⁺ +ⁿ c⁻) ≡⟨ sym (lem0 b⁺ c⁺ a⁻ c⁻) ⟩
+  (b⁺ +ⁿ c⁺) +ⁿ (a⁻ +ⁿ c⁻) ∎)
+
++-respects-relˡ : (a b c : ℕ × ℕ) → rel b c → (a +' b) ≡ (a +' c)
++-respects-relˡ a@(a⁺ , a⁻) b@(b⁺ , b⁻) c@(c⁺ , c⁻) p = eq/ {R = rel} (a⁺ +ⁿ b⁺ , a⁻ +ⁿ b⁻) (a⁺ +ⁿ c⁺ , a⁻ +ⁿ c⁻) (
+  (a⁺ +ⁿ b⁺) +ⁿ (a⁻ +ⁿ c⁻) ≡⟨ lem0 a⁺ b⁺ a⁻ c⁻ ⟩
+  (a⁺ +ⁿ a⁻) +ⁿ (b⁺ +ⁿ c⁻) ≡[ i ]⟨ (a⁺ +ⁿ a⁻) +ⁿ p i ⟩
+  (a⁺ +ⁿ a⁻) +ⁿ (c⁺ +ⁿ b⁻) ≡⟨ sym (lem0 a⁺ c⁺ a⁻ b⁻) ⟩
+  (a⁺ +ⁿ c⁺) +ⁿ (a⁻ +ⁿ b⁻) ∎)
+
+_+''_ : ℤ → ℤ → ℤ
+_+''_ = rec2 {R = rel} {B = ℤ} φ _+'_ +-respects-relʳ +-respects-relˡ
+  where abstract φ = isSetℤ
+
+-- normalization of isSetℤ explodes. Therefore we wrap this with expanded cases
+_+_ : ℤ → ℤ → ℤ
+x@([ _ ])               + y@([ _ ])               = x +'' y
+x@([ _ ])               + y@(eq/ _ _ _ _)         = x +'' y
+x@(eq/ _ _ _ _)         + y@([ _ ])               = x +'' y
+x@(eq/ _ _ _ _)         + y@(eq/ _ _ _ _)         = x +'' y
+x@(eq/ _ _ _ _)         + y@(squash/ a b p q i j) = isSetℤ _ _ (cong (x +_) p) (cong (x +_) q) i j
+x@([ _ ])               + y@(squash/ a b p q i j) = isSetℤ _ _ (cong (x +_) p) (cong (x +_) q) i j
+x@(squash/ a b p q i j) + y                       = isSetℤ _ _ (cong (_+ y) p) (cong (_+ y) q) i j
```

Comment:
This was my question in #456. With the proof done, one could drop this wrapper, again. Should we leave it in?